### PR TITLE
Lots of different fixes.

### PR DESCRIPTION
**Pull request by vonloxley**
Everything was testet against ASE 15.7.

### CODE REVIEW

**Comment by teverett** [comment]
@vonloxley please fix the merge conflicts

**Comment by vonloxley** [comment]
Done, not sure why it’s still marked as conflicting.

**Comment by teverett** [comment]
me either.  maybe try closing this PR and making a new one.  Be sure to do a full pull first.

**Comment by vonloxley** [comment]
I call it a github-bug

**Comment by teverett** [comment]
@KvanTTT we good with this?

File: tsql/ReadMe.md [hunk 1]
```diff
@@ -4,4 +4,11 @@ This is a community supported grammar file for t-sql, it is cool, yet not very c
...
-and bit by bit make it a more complete grammar.
\ No newline at end of file
+and bit by bit make it a more complete grammar.
+
+Use with Antlr 4.7
+------------------
+This grammar uses upper-case key-words only. That means, as sql is case-insensitive, that
+it can’t be used with Antlr 4.7 directly. You need a CharStream that converts to 
+upper-case for the lexer. An implementation for java can be found [on Github](https://github.com/antlr/antlr4/pull/2048/commits/8043e85d178a318172bc94639bd0c131d7e3c60c).
+
```
Comment (KvanTTT) [inline tsql/ReadMe.md]: There is a part of another PR. Is it possible to remove it?

File: tsql/TSqlLexer.g4 [hunk 2]
```diff
@@ -808,7 +816,7 @@ XSINIL:                                'XSINIL';
 DOLLAR_ACTION:                         '$ACTION';
 
 SPACE:              [ \t\r\n]+    -> skip;
-COMMENT:            '/*' .*? '*/' -> channel(HIDDEN);
+COMMENT:            '/*' (COMMENT|.)*? '*/' -> channel(HIDDEN);
```
Comment (KvanTTT) [inline tsql/TSqlLexer.g4]: Could you add a link to nested comment description? https://docs.microsoft.com/en-us/sql/t-sql/language-elements/slash-star-comment-transact-sql
Spaces before and after vertical bar: `COMMENT | .`.

File: tsql/TSqlParser.g4 [hunk 3]
```diff
@@ -3555,155 +3584,424 @@ id
 simple_id
```
Comment (KvanTTT) [inline tsql/TSqlParser.g4]: Good job. How did you find out that these tokens can be also identifiers?

**Comment by teverett** [comment]
So we're good now?

**Comment by KvanTTT** [comment]
Some fixes to be resolved.

File: tsql/TSqlParser.g4 (hunk 3 above)
Comment (vonloxley) [reply tsql/TSqlParser.g4]: We’ve got an rather good list of reserved words, I’m attaching it.
[reserved_words.txt](https://github.com/antlr/grammars-v4/files/1502379/reserved_words.txt)

File: tsql/TSqlLexer.g4 (hunk 2 above)
Comment (vonloxley) [reply tsql/TSqlLexer.g4]: Done.

File: tsql/ReadMe.md (hunk 1 above)
Comment (vonloxley) [reply tsql/ReadMe.md]: Sorry for that. I can rewrite the history of this branch if you prefer.

Comment (KvanTTT) [reply tsql/ReadMe.md]: It would be great.

Comment (vonloxley) [reply tsql/ReadMe.md]: Force-pushed…

**Comment by teverett** [comment]
It looks like this PR resolves issue 754.  The .errors file needs to be removed
```
Parsing :/home/travis/build/antlr/grammars-v4/tsql/examples/pr754.sql
com.khubla.antlr.antlr4test.AssertErrorsException: no errors found, but errors file exists /home/travis/build/antlr/grammars-v4/tsql/examples/pr754.sql.errors

```

File: tsql/TSqlParser.g4 [hunk 4]
```diff
@@ -1950,14 +1959,17 @@ create_statistics
...
 
 // https://msdn.microsoft.com/en-us/library/ms174979.aspx
 create_table
-    : CREATE TABLE table_name '(' column_def_table_constraints ','? ')' table_options* (ON id | DEFAULT)? (TEXTIMAGE_ON id | DEFAULT)?';'?
+    : CREATE TABLE table_name '(' column_def_table_constraints ','? ')' (LOCK simple_id)? table_options* (ON id | DEFAULT)? (TEXTIMAGE_ON id | DEFAULT)?';'?
     ;
 
 table_options
-    : WITH '(' index_option (',' index_option)* ')'
+    : WITH '('? index_option (',' index_option)* ')'?
```
Comment (KvanTTT) [inline tsql/TSqlParser.g4]: Correct record: `WITH ('(' index_option (',' index_option)* ')' | index_option (',' index_option)*)`. Single parenthesis is not correct in lanugage but allowed in your rule.

File: tsql/TSqlParser.g4 [hunk 5]
```diff
@@ -2635,6 +2654,9 @@ go_statement
 use_statement
     : USE database=id ';'?
     ;
+setuser_statement
```
Comment (KvanTTT) [inline tsql/TSqlParser.g4]: Line break.

File: tsql/TSqlParser.g4 [hunk 6]
```diff
@@ -2681,6 +2704,9 @@ column_definition
       ROWGUIDCOL?
       column_constraint*
     ;
+materialized_column_definition
```
Comment (KvanTTT) [inline tsql/TSqlParser.g4]: Line break.

File: tsql/TSqlParser.g4 [hunk 7]
```diff
@@ -2779,7 +2805,7 @@ expression
     | bracket_expression
     | unary_operator_expression
     | expression op=('*' | '/' | '%') expression
-    | expression op=('+' | '-' | '&' | '^' | '|') expression
+    | expression op=('+' | '-' | '&' | '^' | '|' | DOUBLE_BAR) expression
```
Comment (KvanTTT) [inline tsql/TSqlParser.g4]: Why not an ordinary introduced string literal `||`?

File: tsql/TSqlParser.g4 [hunk 8]
```diff
@@ -3211,7 +3237,7 @@ insert_with_table_hints
 // READCOMMITTEDLOCK, READPAST, READUNCOMMITTED, REPEATABLEREAD, ROWLOCK, TABLOCK, TABLOCKX
 // UPDLOCK, XLOCK)
 table_hint
-    : NOEXPAND? ( INDEX '(' index_value (',' index_value)* ')'
+    : NOEXPAND? ( INDEX '('? index_value (',' index_value)* ')'?
```
Comment (KvanTTT) [inline tsql/TSqlParser.g4]: Correct: `('(' index_value (',' index_value)* ')' | index_value (',' index_value)*)`.

**Comment by KvanTTT** [comment]
Overall ok, but some smal issues still need to be resolved.

File: tsql/TSqlParser.g4 [hunk 9]
```diff
@@ -1950,14 +1959,17 @@ create_statistics
       (WITH (FULLSCAN | SAMPLE DECIMAL (PERCENT | ROWS) | STATS_STREAM)
             (',' NORECOMPUTE)? (',' INCREMENTAL EQUAL on_off)? )? ';'?
     ;
+update_statistics
+    : UPDATE (INDEX|ALL)? STATISTICS full_table_name id?  (USING DECIMAL VALUES)?
```
[the remark this answers is not in the document]
Comment (vonloxley) [reply tsql/TSqlParser.g4]: Are you sure? The grammar is rather mixed in this regard:
```
[~/work/grammars-v4/tsql (fixes)]$ grep '[^\ ]|' TSqlParser.g4 |wc -l
138
```

File: tsql/TSqlParser.g4 (hunk 4 above)
Comment (vonloxley) [reply tsql/TSqlParser.g4]: The paranthesis are optional in SYBASE TSQL, which is the dialect I’m using.

File: tsql/TSqlParser.g4 (hunk 7 above)
Comment (vonloxley) [reply tsql/TSqlParser.g4]: I prefer terminals to literals for operators. But if you insist…

File: tsql/TSqlParser.g4 (hunk 9 above)
[the remark this answers is not in the document]
Comment (KvanTTT) [reply tsql/TSqlParser.g4]: In other places it also should be fixed. I notices this one.

File: tsql/TSqlParser.g4 (hunk 4 above)
Comment (KvanTTT) [reply tsql/TSqlParser.g4]: Expression with only one parenthesis is allowed? I'm sure this is incorrect: `WITH '(' index_option (',' index_option)*`. Check my sentence more carefully.

File: tsql/TSqlParser.g4 (hunk 7 above)
Comment (KvanTTT) [reply tsql/TSqlParser.g4]: There are a mix of ordinary literals and identifiers. Please use the certain ones.

Comment (vonloxley) [reply tsql/TSqlParser.g4]: Tried that, got an error: ```cannot create implicit token for string literal in non-combined grammar: '||'```. If you know, how to do it I’d be happy to learn.

Comment (KvanTTT) [reply tsql/TSqlParser.g4]: Strange. Such error occurs if the token is not defined explicitly in Lexer. Did you remove introduced `DOUBLE_BAR` token?

Comment (vonloxley) [reply tsql/TSqlParser.g4]: Yes, I did indeed. Committing now.

**Comment by KvanTTT** [comment]
Thanks for a great job!

**Pull request merged**